demo/views/pages/play-area/wizard-wrapper-stub/wizard-wrapper-stub.js

```diff
@@ -0,0 +1,65 @@
+import Button from 'components/button';
+import React from 'react';
+import Wizard from 'components/wizard';
+
+class WizardWrapperStub extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      steps: [
+        { label: 'foo', state: 'complete' },
+        { label: 'bar', state: 'current' },
+        { label: 'baz', state: 'incomplete' }
+      ],
+      current: 2
+    };
+  }
+
+  setStepState = current => (step, i) => {
+    let state = 'incomplete';
+    if (current === i) { state = 'current'; }
+    if (current > i) { state = 'complete'; }
+    console.log({ state, label: step.label });
+    return { state, label: step.label };
+  };
+
+  previous = () => {
+    let { current, steps } = this.state;
+
+    current = current === 0 ? 0 : current - 1;
+
+    steps = steps.map(this.setStepState(current));
+
+    this.setState({ current, steps });
+  }
+
+  next = () => {
+    let { current } = this.state;
+    let { steps } = this.state;
+
+    current = current === steps.length - 1 ? current : current + 1;
+
+    steps = steps.map(this.setStepState(current));
+
+    this.setState({ current, steps });
+  }
+
+  render() {
+    return (
+      <div className='demo-wizard-wrapper-stub'>
+        <Wizard { ...this.state }>
+          <div>Step 1</div>
+          <div>Step 2</div>
+          <div>Step 3</div>
```

Comment:
I don't think this is particularly intuitive, unless you know how `Wizard` behaves it looks like it is rendering all 3 items at once.

I think it is also particularly restricting as the `Wizard` is responsible for deciding what to render and where it is rendered. It is also heavily dictated by complex props, which need to adhere to a certain format shared down with it's children. These children also have to adhere to the correct index, which may cause issues if you ever want to easily add conditional step etc.

If we instead have developers write this implementation detail in their app it makes it more understandable and flexible to different situations, rather than having to maintain a heavy API that will be prone to change and snowball into even more complexity over time. We have definitely been guilty of this in the past, and it is a pattern we need to break.

```js
render() {
  return (
    <div className='demo-wizard-wrapper-stub'>
      <StepIndicators>
        {
          this.state.steps.map(step => (
            <StepIndicator status={ step.state }>{ step.label }</StepIndicator>
          ));
        }
      </StepIndicators>

      { this.state.current === 1 && <div>Step 1</div> }
      { this.state.current === 2 && <div>Step 2</div> }
      { this.state.current === 3 && <div>Step 3</div> }

      <div className='demo-wizard-wrapper-stub__buttons'>
        <Button onClick={ this.previous }>Previous</Button>
        <Button onClick={ this.next }>Next</Button>
      </div>
    </div>
  );
}
```

While this means there is a bit more code on the implementation side, it is easier to understand exactly what is going on, and can be easily manipulated in various ways if needed with no dictation on the data structure. In reality, this would probably be broken down into a few different components - one handling the indicators and another handling the step rendering.

Reply:
I put the wizard together quickly to demonstrate the StepSequence (same as StepIndicators) which is basically built like you have above

This is it here: https://github.com/Sage/carbon/pull/1866/files?utf8=%E2%9C%93&diff=unified#diff-f76d855afc9eb13f8df6f2d66db69cf8

Reply:
You're suggesting that the Wizard contents are built manually by the dev:

```
<Wizard>
  <StepSequence steps={ steps } />
  { this.state.current === 1 && <div>Step 1</div> }
  { this.state.current === 2 && <div>Step 2</div> }
  { this.state.current === 3 && <div>Step 3</div> }
  <Nav />
</Wizard>
```

rather than

```
<Wizard configuredJSON>
  <div>Step 1</div>
  <div>Step 2</div>
  <div>Step 3</div>
</Wizard>
```

Reply:
I wonder about doing both though,

In the first example, the `Wizard` wrapper seems to have no purpose, the user can put together a StepSequence, ButtonPair and their own content and state implementation and I think all I'd need to deliver would be a step sequence

Why not also provide a Wizard that does it for you for the basic cases, then pull it apart if you need to? It could even provide the state for a quick option for simple wizards

Reply:
From what I have experienced there is never a basic case with something as complex as this. There is your current implementation, but the next squads implementation will be different in some way so they cannot reuse your `Wizard` component - or they need to add their little adjustments to it and it soon escalates into a mess of configurations. Maybe the confusion here is that the `Wizard` you're putting together might better sit in your app as the view layer rather than in Carbon as a reusable component, but even then I think being a little bit more verbose with the code opens up so much more flexibility and readability for future development. You're hiding the readability of React within a JSON config object which is doing some magic behind the scenes.

There is some merit in delivering reusable compositions of components, and it is something we're discussing with UX, but we haven't worked out what they are exactly yet or whether they are a valid thing for Carbon to own. 

Also in your specific case, are you sure the wizard nav even sits outside of the steps? Are they not buttons inside each step which act as an individual form-per-step?